### PR TITLE
precompute EmptyMerklePath roots

### PR DESCRIPTION
**Pull request by LarryRuane**
This is an alternative implementation of #4131.

### CODE REVIEW

File: src/gtest/test_merkletree.cpp [hunk 1]
```diff
@@ -223,25 +221,29 @@ TEST(merkletree, SaplingVectors) {
...
+    std::array<libzcash::SHA256Compress, 65> computed;
 
-    for (size_t depth = 0; depth <= 64; depth++) {
-        expect_test_vector(empty_roots[depth], emptyroots.empty_root(depth));
+    computed.at(0) = libzcash::SHA256Compress::uncommitted();
+    ASSERT_TRUE(emptyroots.empty_root(0) == computed.at(0));
+    for (size_t d = 1; d <= 64; d++) {
+        computed.at(d) = libzcash::SHA256Compress::combine(computed.at(d-1), computed.at(d-1), d-1);
+        ASSERT_TRUE(emptyroots.empty_root(d) == computed.at(d));
     }
```
Comment (daira) [inline src/gtest/test_merkletree.cpp]: Why does each table contain more roots than we'll use?

(I realise that's what we did before this PR, but I don't see any reason for it.)

Comment (str4d) [reply src/gtest/test_merkletree.cpp]: IIRC it was just to exercise a much larger range than we'd likely need. At the time, we didn't know the exact depth we'd need for Sapling.

File: src/zcash/IncrementalMerkleTree.cpp [hunk 2]
```diff
@@ -35,6 +35,392 @@ PedersenHash PedersenHash::uncommitted() {
...
+    uint256(std::vector<unsigned char>{
+        0x60, 0x0e, 0x6f, 0x93, 0xe7, 0x3d, 0x7a, 0xbd,
+        0x4e, 0xe0, 0xa6, 0x5c, 0xb1, 0xb1, 0x9a, 0xa3,
+        0xec, 0xc5, 0x25, 0x68, 0x9d, 0xbf, 0x17, 0x77,
+        0x96, 0x58, 0x74, 0x1b, 0x95, 0xc1, 0x5a, 0x55,
+    }),
+};
+
+PedersenHash PedersenHash::EmptyRoot(size_t depth) {
+    assert(depth <= sizeof(pedersen_empty_roots)/sizeof(pedersen_empty_roots[0]));
```
Comment (str4d) [inline src/zcash/IncrementalMerkleTree.cpp]: Why not `depth < pedersen_empty_roots.size()`?

File: src/zcash/IncrementalMerkleTree.cpp [hunk 3]
```diff
@@ -51,6 +437,410 @@ SHA256Compress SHA256Compress::combine(
...
+    uint256(std::vector<unsigned char>{
+        0x5b, 0x8e, 0x14, 0xfa, 0xca, 0xc8, 0xa7, 0xc7,
+        0xa3, 0xbf, 0xee, 0x8b, 0xae, 0x71, 0xf2, 0xf7,
+        0x79, 0x3d, 0x3a, 0xd5, 0xfe, 0x33, 0x83, 0xf9,
+        0x3a, 0xb6, 0x06, 0x1f, 0x2a, 0x11, 0xbb, 0x02
+    }),
+};
+
+SHA256Compress SHA256Compress::EmptyRoot(size_t depth) {
+    assert(depth <= sizeof(sha256_empty_roots)/sizeof(sha256_empty_roots[0]));
```
Comment (str4d) [inline src/zcash/IncrementalMerkleTree.cpp]: Why not `depth < sha256_empty_roots.size()`?

File: src/zcash/IncrementalMerkleTree.cpp [hunk 4]
```diff
@@ -35,6 +35,392 @@ PedersenHash PedersenHash::uncommitted() {
...
+        0x60, 0x0e, 0x6f, 0x93, 0xe7, 0x3d, 0x7a, 0xbd,
+        0x4e, 0xe0, 0xa6, 0x5c, 0xb1, 0xb1, 0x9a, 0xa3,
+        0xec, 0xc5, 0x25, 0x68, 0x9d, 0xbf, 0x17, 0x77,
+        0x96, 0x58, 0x74, 0x1b, 0x95, 0xc1, 0x5a, 0x55,
+    }),
+};
+
+PedersenHash PedersenHash::EmptyRoot(size_t depth) {
+    assert(depth <= sizeof(pedersen_empty_roots)/sizeof(pedersen_empty_roots[0]));
+    return pedersen_empty_roots[depth];
```
Comment (str4d) [inline src/zcash/IncrementalMerkleTree.cpp]: This alters the current behaviour. In the previous implementation `array::at` was used, resulting in an error being thrown if the index was out of bounds. Here we assert and then do an unchecked lookup, meaning that if assertions are disabled by someone, the out-of-bounds read will go through. Please replace these two lines with:
```suggestion
    return pedersen_empty_roots.at(depth);
```

File: src/zcash/IncrementalMerkleTree.cpp [hunk 5]
```diff
@@ -51,6 +437,410 @@ SHA256Compress SHA256Compress::combine(
...
+        0x5b, 0x8e, 0x14, 0xfa, 0xca, 0xc8, 0xa7, 0xc7,
+        0xa3, 0xbf, 0xee, 0x8b, 0xae, 0x71, 0xf2, 0xf7,
+        0x79, 0x3d, 0x3a, 0xd5, 0xfe, 0x33, 0x83, 0xf9,
+        0x3a, 0xb6, 0x06, 0x1f, 0x2a, 0x11, 0xbb, 0x02
+    }),
+};
+
+SHA256Compress SHA256Compress::EmptyRoot(size_t depth) {
+    assert(depth <= sizeof(sha256_empty_roots)/sizeof(sha256_empty_roots[0]));
+    return sha256_empty_roots[depth];
```
Comment (str4d) [inline src/zcash/IncrementalMerkleTree.cpp]: This alters the current behaviour. In the previous implementation `array::at` was used, resulting in an error being thrown if the index was out of bounds. Here we assert and then do an unchecked lookup, meaning that if assertions are disabled by someone, the out-of-bounds read will go through. Please replace these two lines with:
```suggestion
    return sha256_empty_roots.at(depth);
```

File: src/zcash/IncrementalMerkleTree.cpp (hunk 2 above)
Comment (daira) [reply src/zcash/IncrementalMerkleTree.cpp]:
```suggestion
    assert(depth < pedersen_empty_roots.size());
```

File: src/zcash/IncrementalMerkleTree.cpp (hunk 3 above)
Comment (daira) [reply src/zcash/IncrementalMerkleTree.cpp]:
```suggestion
    assert(depth < sha256_empty_roots.size());
```

**Comment by daira** [comment]
+1 to @str4d's suggested change. Ping @LarryRuane

File: src/zcash/IncrementalMerkleTree.cpp (hunk 2 above)
Comment (daira) [reply src/zcash/IncrementalMerkleTree.cpp]:
```suggestion
```

File: src/zcash/IncrementalMerkleTree.cpp (hunk 3 above)
Comment (daira) [reply src/zcash/IncrementalMerkleTree.cpp]:
```suggestion
```

File: src/zcash/IncrementalMerkleTree.cpp (hunk 2 above)
Comment (LarryRuane) [reply src/zcash/IncrementalMerkleTree.cpp]: Better to remove the assert?

**Comment by LarryRuane** [comment]
Latest commit for review suggestions, rebased.

**Comment by zebambam** [comment]
Hi, I believe that I started the original implementation of this without giving a ticket number, or at least I couldn't find one when I went looking just now. Sorry about that. Here's one to cover it. #4155

**Comment by LarryRuane** [comment]
@str4d, the good thing is that if the hard-coded empty roots are wrong, the unit test doesn't pass. (I actually modified one of the values, and verified that the test fails.)

**Comment by LarryRuane** [comment]
Just to confirm that @zebambam was addressing the correct cause of the problem addressed by both his PR the current PR, I added a call to `assert(0)` at the top of `src/bitcoind.cpp: main()`, which occurs after all of the static initialization. I ran `time src/zcashd`. Without this PR, a typical result (varied slightly across multiple runs) is:
```
$ time src/zcashd
zcashd: bitcoind.cpp:192: int main(int, char**): Assertion `0' failed.
Aborted (core dumped)

real	0m0.213s
user	0m0.074s
sys	0m0.004s
```
With this PR:
```
real	0m0.123s
user	0m0.004s
sys	0m0.000s
```
(In this case, the user time was always exactly the same, 0.004 seconds.) So this seems as expected, if this is a good way to demonstrate the effect.

**Comment by Eirik0** [comment]
@zkbot r+

**Comment by Eirik0** [comment]
Looks like a combination of timeouts and failures which do not appear related to this PR. @zkbot retry

**Pull request merged**